contracts/PooledStaking.sol

```diff
@@ -0,0 +1,512 @@
+
+    // transfer tokens from specified contract to us
+    // token transfer should be approved by the specified address
+    Vault.deposit(token, from, amount);
+
+    rewards[rewardCount] = Reward(amount, now, contractAddress, 0); // add new reward
+
+    // TODO: recheck this
+    // similar to pushBurn
+    if (rewardCount != firstReward && rewards[rewardCount - 1].rewardedAt > 0) {
+      rewards[rewardCount - 1].next = rewardCount; // set previous reward's next to current id
+    } else {
+      firstReward = rewardCount;
+    }
+
+    ++rewardCount; // update counter
+  }
+
+  function processPendingActions() public {
+    while (true) {
+      Burn storage burn = burns[firstBurn];
+      Deallocation storage deallocation = deallocations[firstDeallocation];
+
+      bool canBurn = burn.burnedAt != 0; // end not reached
+      bool canDeallocate = deallocation.deallocateAt != 0 && deallocation.deallocateAt <= now;
+
+      if (!canBurn && !canDeallocate) {
+        // everything is processed
+        break;
+      }
+
+      // TODO: check if there's enough gas for the next iteration
+
+      // no burn to compare to or deallocation should happen before burn
+      if (!canBurn || deallocation.deallocateAt <= burn.burnedAt) {
+        _processFirstDeallocation();
+        continue;
+      }
+
+      _processFirstBurn();
+    }
+  }
+
+  function _processFirstDeallocation() internal {
+    Deallocation storage deallocation = deallocations[firstDeallocation];
+    Staker storage staker = stakers[deallocation.stakerAddress];
+
+    address contractAddress = deallocation.contractAddress;
+    staker.allocations[contractAddress].sub(deallocation.amount);
+    staker.pendingDeallocations[contractAddress].add(deallocation.amount);
+
+    uint nextDeallocation = deallocation.next;
+    delete deallocations[firstDeallocation];
+    firstDeallocation = nextDeallocation;
+  }
+
+  function _processFirstBurn() internal {
+
+    Burn storage burn = burns[firstBurn];
+    Contract storage _contract = contracts[burn.contractAddress];
+    address contractAddress = burn.contractAddress;
+
+    uint burnAmount = burn.amount <= _contract.staked ? burn.amount : _contract.staked;
+    uint burned = 0;
+    uint newContractStake = 0;
+
+    for (uint i = 0; i < _contract.stakers.length; i++) {
+
+      Staker storage staker = stakers[_contract.stakers[i]];
+
+      uint oldAllocation = staker.allocations[contractAddress];
+
+      // formula: staker_burn = staker_allocation / total_contract_stake * contract_burn
+      // reordered for precision loss prevention
+      uint stakerBurn = oldAllocation.mul(burnAmount).div(_contract.staked);
+      uint newStake = staker.staked.sub(stakerBurn);
+      staker.allocations[contractAddress] = oldAllocation.sub(stakerBurn);
+
+      // reduce other contracts' stakes if needed
+      for (uint j = 0; j < staker.contracts.length; j++) {
```

Comment:
since this is O(m*n) we need tests around gas usage here to see exactly what can we support and potentially make a test fail if a certain amount of gas is used that is not acceptable